projects/baby_driver/test/test_gpio_get.c

```diff
@@ -0,0 +1,178 @@
+
+  s_callback_counter = 0;
+  s_callback_counter_state = 0;
+  memset(s_received_data, 0, sizeof(s_received_data));
+  memset(s_received_data_state, 0, sizeof(s_received_data_state));
+  TEST_ASSERT_OK(dispatcher_register_callback(BABYDRIVER_MESSAGE_STATUS,  //
+                                              prv_callback_gpio_get,      //
+                                              NULL));                     //
+
+  TEST_ASSERT_OK(dispatcher_register_callback(BABYDRIVER_MESSAGE_GPIO_GET_DATA,  //
+                                              prv_callback_gpio_get_state,       //
+                                              NULL));                            //
+}
+
+void teardown_test(void) {}
+
+void test_invalid_input() {
+  // test for invalid port and valid pin
+  CAN_TRANSMIT_BABYDRIVER(BABYDRIVER_MESSAGE_GPIO_GET_COMMAND,  //
+                          INVALID_PORT,                         //
+                          VALID_PIN, 0, 0, 0, 0, 0);            //
+
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+
+  TEST_ASSERT_EQUAL(1, s_callback_counter);
+  TEST_ASSERT_EQUAL(BABYDRIVER_MESSAGE_STATUS, s_received_data[0]);
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, s_received_data[1]);
+
+  // test for invalid pin and valid port
+  CAN_TRANSMIT_BABYDRIVER(BABYDRIVER_MESSAGE_GPIO_GET_COMMAND,  //
+                          VALID_PORT,                           //
+                          INVALID_PIN, 0, 0, 0, 0, 0);          //
+
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+
+  TEST_ASSERT_EQUAL(2, s_callback_counter);
+  TEST_ASSERT_EQUAL(BABYDRIVER_MESSAGE_STATUS, s_received_data[0]);
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, s_received_data[1]);
+
+  // test for invalid pin and port
+  CAN_TRANSMIT_BABYDRIVER(BABYDRIVER_MESSAGE_GPIO_GET_COMMAND,  //
+                          INVALID_PORT,                         //
+                          INVALID_PIN, 0, 0, 0, 0, 0);          //
+
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+
+  TEST_ASSERT_EQUAL(3, s_callback_counter);
+  TEST_ASSERT_EQUAL(BABYDRIVER_MESSAGE_STATUS, s_received_data[0]);
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, s_received_data[1]);
+}
+
+// Testing that gpio_get works by checking if it outputs the correct state of a given pin and port
+void test_gpio_get_valid_input() {
+  // Valid pin and port with a high state
+  GpioSettings settings = {
+    .direction = GPIO_DIR_OUT,        //
+    .state = GPIO_STATE_HIGH,         //
+    .resistor = GPIO_RES_NONE,        //
+    .alt_function = GPIO_ALTFN_NONE,  //
+  };
+
+  GpioAddress address = {
+    .port = VALID_PORT,  //
+    .pin = VALID_PIN,    //
+  };
+
+  TEST_ASSERT_OK(gpio_init_pin(&address, &settings));
```

Comment:
Okay, so here's the deal: on x86, if a pin is initialized as an input, the `gpio` library will return values from an uninitialized array you can't modify in `gpio_get_state` - this is a design flaw with the library. So, with the initialization corrected above, you won't be able to modify the state of the gpio pin.

However, to save the day are what is called *mocks*. When you mock a function, you write an alternate implementation of that function which is called instead of the original function within a test. So, I would suggest that you mock `gpio_get_state`, so that you can return a value of your choosing instead of relying on the whims of the `gpio` library. We have some macro magic for mocks, so it'll look like this:
```c
StatusCode TEST_MOCK(gpio_get_state)(const GpioAddress *address, GpioState *state) {
  // your alternate implementation of gpio_get_state...
  *state = s_returned_state; // for example
  return STATUS_CODE_OK;
}
```
Then you just have to add the following line to projects/baby_driver/rules.mk to inform the build system about your mock:
```makefile
$(T)_adc_read_MOCKS := gpio_get_state
```
Then instead of actually initializing a GPIO pin to set the state returned by `gpio_get_state`, you can just set a static variable, avoiding the whole initialization problem with the `gpio` library.